### DISCUSSION TITLE
Unexpected partial matches for simple quoted symbol rule.

### DISCUSSION THREAD

**Discussion by cpu** [tree-sitter/tree-sitter#1873] [Q&A]
Hello!

I'm very new to writing grammars and I'm presently stumped on a problem and would love some help.

My objective is to match "quoted symbol" input like `'a` or `'someFunc200` where there's a leading `'` literal and then something like an identifier afterwards. 

The problem I'm having is that a well-formed symbol with an invalid trailing `'` is being matched as two symbols, one of which has no characters after the `'`. I believe this is causing me grief as I try to write a rule for character literals.

e.g. parsing this input:
```
'a'
```

produces this tree:
```lisp
(source_file [0, 0] - [1, 0]
  (symbol [0, 0] - [0, 2])
  (symbol [0, 2] - [0, 3]))
```

when I actually want it to error, and produce something like:
```lisp
(source_file [0, 0] - [1, 0]
  (symbol [0, 0] - [0, 2])
  <error here!>
```

Here's a minimal version of my grammar.
```js
module.exports = grammar({
  name: 'test',

  rules: {
    source_file: $ => repeat($.symbol),

    symbol: $ => seq(
      "'",
      token.immediate(/[a-zA-Z_\$]\w*/),
    ),
  }
});
```

Can anyone help me understand why the trailing `'` is parsed as a `symbol` when the `seq` rule in `symbol` should have failed to match the regexp in the `token.immediate` invocation?

Many thanks!

**Comment by ahelwer** [comment]
What does the parser debug output show

**Reply by cpu** [reply]
@ahelwer Thanks for the reply. I had missed the presence of the debug flags in `tree-sitter parse`, very helpful! 

Here's what it spits out:
```
new_parse
process version:0, version_count:1, state:1, row:0, col:0
lex_internal state:0, row:0, column:0
  consume character:'''
lexed_lookahead sym:', size:1
shift state:5
process version:0, version_count:1, state:5, row:0, col:1
lex_internal state:0, row:0, column:1
  consume character:'a'
lexed_lookahead sym:symbol_token1, size:1
shift state:4
process version:0, version_count:1, state:4, row:0, col:2
lex_internal state:0, row:0, column:2
  consume character:'''
lexed_lookahead sym:', size:1
reduce sym:symbol, child_count:2
shift state:5
process version:0, version_count:1, state:5, row:0, col:3
lex_internal state:0, row:0, column:3
  skip character:10
lexed_lookahead sym:end, size:1
detect_error
resume version:0
recover_with_missing symbol:symbol_token1, state:2
recover_to_previous state:1, depth:3
recover_eof
process version:1, version_count:3, state:2, row:0, col:3
reduce sym:source_file, child_count:1
accept
select_smaller_error symbol:source_file, over_symbol:ERROR
process version:2, version_count:3, state:1, row:0, col:3
reduce sym:source_file, child_count:0
accept
select_smaller_error symbol:source_file, over_symbol:source_file
done
(source_file [0, 0] - [1, 0]
  (symbol [0, 0] - [0, 2])
  (symbol [0, 2] - [0, 3]))
 ```

**Reply by ahelwer** [reply]
So it does have a parse error, it just recovers. Useful for highlighting but less useful if you want to surface parse errors to your user.
